x-pack/legacy/plugins/maps/public/layers/sources/es_geo_grid_source/es_geo_grid_source.js

```diff
@@ -175,33 +164,132 @@ export class ESGeoGridSource extends AbstractESAggSource {
-      aggSchemas.all
-    );
-    searchSource.setField('aggs', aggConfigs.toDsl());
-    const esResponse = await this._runEsQuery({
-      requestId: this.getId(),
-      requestName: layerName,
-      searchSource,
-      registerCancelCallback,
-      requestDescription: i18n.translate('xpack.maps.source.esGrid.inspectorDescription', {
-        defaultMessage: 'Elasticsearch geo grid aggregation request',
-      }),
-    });
 
-    const tabifiedResp = tabifyAggResponse(aggConfigs, esResponse);
-    const { featureCollection } = convertToGeoJson({
-      table: tabifiedResp,
-      renderAs: this._descriptor.requestType,
+    let bucketsPerGrid = 1;
+    this.getMetricFields().forEach(metricField => {
+      if (metricField.getAggType() === AGG_TYPE.TERMS) {
+        // each terms aggregation increases the overall number of buckets per grid
+        bucketsPerGrid++;
+      }
     });
 
+    const features = [];
+    // Do not use composite aggregation when there are no terms sub-aggregations
+    // see https://github.com/elastic/kibana/pull/57875#issuecomment-590515482 for explanation on using separate code paths
+    if (bucketsPerGrid === 1) {
+      searchSource.setField('aggs', {
+        gridSplit: {
+          geotile_grid: {
+            field: this._descriptor.geoField,
+            precision: searchFilters.geogridPrecision,
+          },
+          aggs: {
+            gridCentroid: {
+              geo_centroid: {
+                field: this._descriptor.geoField,
+              },
+            },
+            ...this.getValueAggsDsl(indexPattern),
+          },
+        },
+      });
+
+      const esResponse = await this._runEsQuery({
+        requestId: this.getId(),
+        requestName: layerName,
+        searchSource,
+        registerCancelCallback,
+        requestDescription: i18n.translate('xpack.maps.source.esGrid.inspectorDescription', {
+          defaultMessage: 'Elasticsearch geo grid aggregation request',
+        }),
+      });
+
+      features.push(...convertRegularRespToGeoJson(esResponse, this._descriptor.requestType));
+    } else {
+      const gridsPerRequest = Math.floor(DEFAULT_MAX_BUCKETS_LIMIT / bucketsPerGrid);
+      const aggs = {
+        compositeSplit: {
+          composite: {
+            size: gridsPerRequest,
+            sources: [
+              {
+                gridSplit: {
+                  geotile_grid: {
+                    field: this._descriptor.geoField,
+                    precision: searchFilters.geogridPrecision,
+                  },
+                },
+              },
+            ],
+          },
+          aggs: {
+            gridCentroid: {
+              geo_centroid: {
+                field: this._descriptor.geoField,
+              },
+            },
+            ...this.getValueAggsDsl(indexPattern),
+          },
+        },
+      };
+
+      let requestCount = 0;
+      let afterKey = null;
+      while (true) {
```

Comment:
`while(true)` is really scary looking :)

Maybe use a do-while loop here with `esResponse.aggregations.compositeSplit.buckets.length < gridsPerRequest` as the stop condition.

Reply:
I do not feel the same way about while(true). I think this reads find and has a circuit breaker. I would prefer to keep as is